W-20283303-email-connector-updates-kt (#3130)
mulesoft/docs-connectors: email/1.8/modules/ROOT/pages/email-send.adoc

Configure *Set Variable* components to define multiple media type attachments, and send these attachments using the Email Connector *Send* operation. Then log the result message using a *Logger* component.

* <<send-attachments-from-api,Send Attachments Received from a REST API>> +
Accept a PDF, MS Word document, and JSON file (for example, an incident report) via multipart REST API, and forward them to a support function by email.

* <<send-sftp-files-by-email,Send Files from an SFTP Server by Email>> +
Poll or list files on an SFTP server and send them as email attachments to a function or mailbox.

[[understanding-attachments]]
== Understand Attachments

This section explains when to use fixed vs dynamic attachments, and what the *Attachments* parameter expects.

=== Using Fixed vs Dynamic Attachments

* Fixed attachments +
These are useful when the same file is always sent with certain emails, for example:
** A legal disclaimer or terms document (PDF) attached to every outgoing contract email
** A cover letter or template (for example, a Word or PDF) that is the same for a batch of notifications
** A static report template or form that recipients always need

> 

Reviewer: ```suggestion
+
```

In these cases you typically read the file once (for example, from classpath or a fixed path) or set it in a variable, then reference it in the *Attachments* map.

* Dynamic attachments +
These are built from runtime data: payload from a flow (File read, HTTP multipart, SFTP read), variables set earlier in the flow, or DataWeave expressions that produce binary or string content. The examples in this page show both patterns.

=== Understanding Attachments Expectations

> The *Attachments* parameter of the *Send* operation expects a DataWeave expression that evaluates to an object (a map). It isn't a Java type nor raw JSON configured in the UI; you write a DataWeave expression (for example in the XML `email:attachments` element or in the Studio *Attachments* field) that produces the map.

Reviewer: ```suggestion
The *Attachments* parameter of the *Send* operation expects a DataWeave expression that evaluates to an object (a map). It isn't a Java type nor raw JSON configured in the UI; you write a DataWeave expression (for example, in the XML `email:attachments` element or in the Studio *Attachments* field) that produces the map.
```

* Map structure: Each key is the **attachment filename** (as shown in the email client). Each value is the attachment content (Binary, String, or a stream).

> * Content type: The map itself does not have a single content type. Each value is the raw content; the connector uses the content as-is. When the value comes from a Mule message (for example, `payload` or `vars.x`) that has a MIME type (for example from an HTTP multipart part or a File read), that type is used for the attachment. For inline DataWeave content you can control the type by using `read()` with a MIME type or by ensuring the value is the correct type (for example, `output application/octet-stream` for binary).

Reviewer: ```suggestion
* Content type: The map itself does not have a single content type. Each value is the raw content; the connector uses the content as-is. When the value comes from a Mule message (for example, `payload` or `vars.x`) that has a MIME type (for example, from an HTTP multipart part or a File read), that type is used for the attachment. For inline DataWeave content, you can control the type by using `read()` with a MIME type or by ensuring the value is the correct type (for example, `output application/octet-stream` for binary).
```

</mule>
----

[[truststore-keystore]]
== Configuring TLS: Truststore and Keystore

When you use SMTPS Connection, you must configure Transport Layer Security (TLS) encryption. The *Trust Store* and optionally *Key Store* settings tell the connector which certificates to use. 

=== Locating the Path

The *Path* in *Trust Store* and *Key Store* setting is resolved relative to (in order):

> * The current classpath (for example files under `src/main/resources` in your project)

Reviewer: ```suggestion
* The current classpath (for example, files under `src/main/resources` in your project)
```

* The file system (relative to the app runtime or absolute path)

Typical locations:

* Project resource: `src/main/resources/truststore.jks` — use path `truststore.jks` or `classpath:truststore.jks`
* Absolute path: `/etc/mule/truststore.jks` (ensure the runtime has read access)

=== Configuring Trust Store

The *Trust Store* holds certificates of servers (or CAs) that your app trusts. For SMTPS, the client uses it to verify the mail server’s certificate.

For SMTPS you almost always need a truststore so the connector can validate the server’s certificate (unless you use *Insecure* for development only).

> * If your server uses a public CA (for example Gmail, Office 365): You can use the JVM’s default trust store (often not required to set one) or export it and point *Path* to that file. Some setups use a custom truststore that contains only the CA or server cert they need.

Reviewer: ```suggestion
* If your server uses a public CA (for example Gmail, Office 365): You can use the JVM’s default truststore (often not required to set one) or export it and point *Path* to that file. Some setups use a custom truststore that contains only the CA or server cert they need.
```

* Creating or updating a truststore with keytool (Java’s `keytool` is usually under `JAVA_HOME/bin`):

> * Import a CA or server certificate into a new trust store:

Reviewer: ```suggestion
* Import a CA or server certificate into a new truststore:
```

+
[source,bash,linenums]
----
keytool -importcert -alias smtpserver -file server.cer -keystore truststore.jks -storepass changeit -noprompt
----
* Replace `server.cer` with the path to the server’s certificate or the CA certificate, and `truststore.jks` with the path where you want the truststore file. Use a strong password in production instead of `changeit`.
* In the Email Connector TLS configuration, set *Path* to the path of this trust store and *Password* to the truststore password.

=== Configuring Keystore

The *Key Store* holds your app’s private key and certificate (client identity). Use it when the SMTPS server requires client certificate authentication (mutual TLS).

> Use it only if the mail server is configured to require a client certificate. Many SMTPS servers (for example, Gmail with user or password) don't require a keystore; they only need the trust store to verify the server.

Reviewer: ```suggestion
Use it only if the mail server is configured to require a client certificate. Many SMTPS servers (for example, Gmail with user or password) don't require a keystore; they only need the truststore to verify the server.
```

</mule>
----

[[send-attachments-from-api]]
== Send Attachments Received from a REST API

> A common use case is accepting files (for example a PDF, an MS Word document, and a JSON file describing an incident) through a REST API and forwarding them to a support function by email. The client sends a `multipart/form-data` POST request; each part is a file or a field (for example, an incident description as JSON).

Reviewer: ```suggestion
A common use case is accepting files (for example, a PDF, an MS Word document, and a JSON file describing an incident) through a REST API and forwarding them to a support function by email. The client sends a `multipart/form-data` POST request; each part is a file or a field (for example, an incident description as JSON).
```

> * Flow outline

Reviewer: ```suggestion
* Flow outline +
```

An HTTP *Listener* (path for example, `/api/support/ticket`) receives the multipart request. The payload is a multipart message; use `payload.parts.<name>` to access each part. Build an attachments map from the parts (using each part’s content and optionally filename from `Content-Disposition`), then call Email *Send* with that map and the desired *To* address (for example, support@company.com).

> * Attachments map from multipart

Reviewer: ```suggestion
* Attachments map from multipart +
```

Each part’s content is available as `payload.parts.<partName>.content`. The filename (if present) is in `payload.parts.<partName>.headers.'Content-Disposition'.filename`. Use a DataWeave expression to build a map: keys = attachment names (for example, original filename or a fixed name), values = part content. The connector uses the part’s content type when available. Binary parts (PDF, PNG, Word, and so on) are already Binary; the connector applies Content-Transfer-Encoding (default Base64) for transport, so you do not need to encode them yourself.

> === Example: PDF, Word, and JSON incident to support

Reviewer: ```suggestion
=== Example: PDF, Word, and JSON Incident to Support
```

					(payload.parts.documentPdf.headers.'Content-Disposition'.filename default 'document.pdf') : payload.parts.documentPdf.content,
					(payload.parts.documentWord.headers.'Content-Disposition'.filename default 'document.docx') : payload.parts.documentWord.content,
					(payload.parts.incidentJson.headers.'Content-Disposition'.filename default 'incident.json') : payload.parts.incidentJson.content
				}
			]]]></email:attachments>
		</email:send>
		<set-payload value="#['Ticket received']" />
	</flow>
</mule>
----

The client must POST to `/api/support/ticket` with `Content-Type: multipart/form-data` and three parts named `documentPdf`, `documentWord`, and `incidentJson`. Each part’s content type (for example, `application/pdf`, `application/vnd.openxmlformats-officedocument.wordprocessingml.document`, `application/json`) is preserved from the request and used for the corresponding email attachment.

[[send-sftp-files-by-email]]
== Send Files from an SFTP Server by Email

You may need to scan an SFTP directory and send the found files as email attachments to a function or mailbox (for example a shared inbox or a processing service). There are two common patterns you can follow:

> * Use an *On New or Updated File* (SFTP connector) to trigger a flow when a new or updated file appears; read that file and send it (or a batch) by email.

Reviewer: ```suggestion
* Use an *On New or Updated File* (SFTP Connector) to trigger a flow when a new or updated file appears; read that file and send it (or a batch) by email.
```

> * Use a *Scheduler* and SFTP *List* to poll a directory, then for each file (or for the whole list) read the file(s) and send one or more emails with attachments.

Reviewer: ```suggestion
* Use a *Scheduler* and SFTP *List* to poll a directory, then for each file (or for the whole list), read the file(s) and send one or more emails with attachments.
```

In both cases, build the *Attachments* map from the file content. For a single file, the map can be `{ filename : payload }` after an SFTP *Read*. For multiple files, collect them (for example, in a *For Each* with *Set Variable* or a DataWeave transform) into a map of filename and content and pass that map to *Attachments*.

> === Example: Poll SFTP directory and email files as attachments

Reviewer: ```suggestion
=== Example: Poll SFTP Directory and Email Files as Attachments
```

		<sftp:connection host="${sftp.host}" port="${sftp.port}" username="${sftp.user}" password="${sftp.password}" />
	</sftp:config>
	<flow name="sftp-to-email-flow">
		<scheduler>
			<scheduling-strategy><fixed-frequency frequency="60" timeUnit="SECONDS" /></scheduling-strategy>
		</scheduler>
		<sftp:list config-ref="SFTP_Config" directoryPath="${sftp.directoryPath}" />
		<foreach>
			<choice>
				<when expression="#[!attributes.directory]">
					<set-variable variableName="filePath" value="#[payload]" />
					<sftp:read config-ref="SFTP_Config" path="#[payload]" />
					<email:send config-ref="Email_SMTP" fromAddress="noreply@example.com" subject="File from SFTP: #[vars.filePath]">
						<email:to-addresses>
							<email:to-address value="${email.recipient}" />
						</email:to-addresses>
						<email:body contentType="text/plain">
							<email:content><![CDATA[#["File received from SFTP: " ++ vars.filePath]]]></email:content>
						</email:body>
						<email:attachments><![CDATA[#[( (vars.filePath splitBy "/")[-1] default 'attachment' ) : payload]]]></email:attachments>
					</email:send>
				</when>
			</choice>
		</foreach>
	</flow>
</mule>
----

> To send all listed files in a single email, first iterate to read each file and collect name and content into a variable (for examplem using *Set Variable* with a DataWeave expression that merges the current map with the new attachment), then call *Send* once with that variable as the attachments map. The previous pattern sends one email per file for simplicity.

Reviewer: ```suggestion
To send all listed files in a single email, first iterate to read each file and collect the name and content into a variable (for example, using *Set Variable* with a DataWeave expression that merges the current map with the new attachment), then call *Send* once with that variable as the attachments map. The previous pattern sends one email per file for simplicity.
```
